Ray Label Based Scheduling REP (#60)
ray-project/enhancements: reps/2025-03-18-label-based-scheduling.md

> ## Summary

Reviewer: Can you link to the old REP here and say this supersedes it?

Author: Done in [4709064](https://github.com/ray-project/enhancements/pull/60/commits/4709064c484a39c0048672f3b47777df49728366)

### General Motivation

> This REP summarizes the current state of the node label scheduling feature enhancement and the remaining work to fully support label based scheduling for Tasks and Actors in Ray.

Reviewer: ```suggestion
This REP summarizes the current state of the node label scheduling feature enhancement and the remaining work to fully support scheduling using label selectors in Ray.
```

Let's center on the term "label selector" rather than "label-based scheduling."

### Should this change be within `ray` or outside?

The change should be within Ray since it's a direct enhancement to the Ray scheduler.

## Stewardship
### Required Reviewers
@MengjinYan

### Shepherd of the Proposal (should be a senior committer)
@edoakes

## Design and Architecture

> ### Current implementation state

Author: Yeah that's done here: https://github.com/ray-project/ray/blob/ffd7cc6a63d9e02fb2fa6aa052568ee33cd6469c/python/ray/remote_function.py#L462,, I'll make it more clear like in https://github.com/ray-project/enhancements/pull/60#discussion_r2006447522

> ### Current implementation state

Author: [ed882e2](https://github.com/ray-project/enhancements/pull/60/commits/ed882e220ac3c02ef5a7f5ccc66312a2342d072c)

> Ray currently supports passing labels to a node through `ray start` with the `--labels` flag in Python, parsing labels from a json string with `parse_node_labels_json`, and associates a list of labels with a Node through the `_labels` field. Node information, including labels, are saved in the `GcsNodeInfo` data struct when a node is added. Ray also supports setting default labels on node add, but currently only sets `ray.io/node-id`.

Author: It's from the task/actor options: https://github.com/ray-project/ray/blob/ffd7cc6a63d9e02fb2fa6aa052568ee33cd6469c/python/ray/remote_function.py#L242

> Ray currently supports passing labels to a node through `ray start` with the `--labels` flag in Python, parsing labels from a json string with `parse_node_labels_json`, and associates a list of labels with a Node through the `_labels` field. Node information, including labels, are saved in the `GcsNodeInfo` data struct when a node is added. Ray also supports setting default labels on node add, but currently only sets `ray.io/node-id`.

Author: ed882e2

> Ray currently supports passing labels to a node through `ray start` with the `--labels` flag in Python, parsing labels from a json string with `parse_node_labels_json`, and associates a list of labels with a Node through the `_labels` field. Node information, including labels, are saved in the `GcsNodeInfo` data struct when a node is added. Ray also supports setting default labels on node add, but currently only sets `ray.io/node-id`.

Reviewer 2: Note that the `_labels` field is not intended for the label based scheduling and was added later in this PR: https://github.com/ray-project/ray/pull/48715

The specification of the labels are actually defined within the `NodeLabelSchedulingStrategy` itself (https://github.com/ray-project/ray/blob/dfa2cb772bb984d1a07601f4a7793f41ec38bc41/python/ray/util/scheduling_strategies.py#L139)

> Ray currently supports passing labels to a node through `ray start` with the `--labels` flag in Python, parsing labels from a json string with `parse_node_labels_json`, and associates a list of labels with a Node through the `_labels` field. Node information, including labels, are saved in the `GcsNodeInfo` data struct when a node is added. Ray also supports setting default labels on node add, but currently only sets `ray.io/node-id`.

Author: Ah I see, that makes sense. I was thinking the labels passed into `_labels` could be used for scheduling but I see that the use case is different and we still need the `label_selector` argument. I fixed the wording in 4f52141 and added a task/actor options example in [ecc2967](https://github.com/ray-project/enhancements/pull/60/commits/ecc2967090e850660f12be92b28b0ede3d5d649c).

To pass labels to a Ray node:
```sh
ray start --head --labels='{"gpu_type": "A100", "region": "us"}'

> ```

Author: 42f3524

    const Node &node, const rpc::LabelMatchExpression &expression) const {
    # Returns a bool based on whether a node's labels satisfy the given condition.
    # Supports exists, not exists, in, and not in conditions. We should also extend 
    # support to equal and not equal.
}
```

### Brief idea
In order to implement full label based scheduling as described in the [public proposal](https://docs.google.com/document/d/1DKhPuZERlLbsU4TIQZZ6POCsm1pVMBgN_yn5r0OmaDI), there are several required changes to the existing API and internal implementation in Ray core. Since most of the core functionality for adding, storing, and retrieving node labels is already implemented, the primary changes proposed here are to update the APIs, support autoscaling, and directly schedule nodes based on label selectors when provided, rather than passing a separate scheduling policy.

### API Design

To pass labels to a Ray node, we will amend the `--labels` argument to `ray start` to accept a string list of key-value pairs. Currently the labels argument accepts a json struct.
```sh
ray start --labels "key1=val1,key2=val2"
```

We will also support sourcing labels from a file using bash:
```sh

> ray start --labels $(cat labels.txt)

Author: [2a9f044](https://github.com/ray-project/enhancements/pull/60/commits/2a9f044077e5f68161d06c6ce24be93180d7de79)

```

> To then schedule based on these labels, we will support passing a `label_selector` argument to the `@ray.remote` decorator.

Reviewer: Note here the motivation: it makes the API composable with other scheduling strategies

Author: [4709064](https://github.com/ray-project/enhancements/pull/60/commits/4709064c484a39c0048672f3b47777df49728366)

```python
@ray.remote(label_selector={"ray.io/accelerator-type": "nvidia-h100"})
class Actor:
    pass
...

@ray.remote(label_selector={"ray.io/market-type": "spot"})
def my_task():
    pass
```

> To schedule placement groups based on labels we will implement support for applying label selectors to placement groups on a per-bundle level. This would require adding a `bundle_label_selector` to the `ray.util.placement_group` constructor.

Author: [ed882e2](https://github.com/ray-project/enhancements/pull/60/commits/ed882e220ac3c02ef5a7f5ccc66312a2342d072c)

```python
# Same labels on all bundles
ray.util.placement_group(
    bundles=[{"GPU": 1}, {"GPU": 1}],

>     bundle_label_selector=[{"ray.io/availability-zone": "us-west4-a"} * 2],

Reviewer: ```suggestion
    bundle_label_selector=[{"ray.io/availability-zone": "us-west4-a"}] * 2,
```

)

> # Different bundles require different labels

Reviewer: ```suggestion
# Different bundles requiring different labels
```

ray.util.placement_group(

>     bundles=[{"CPU": 1}] + [{"GPU": 1} * 2],

Reviewer: ```suggestion
    bundles=[{"CPU": 1}] + [{"GPU": 1}] * 2,
```

Reviewer: please update others as well

    bundle_label_selector=[{"ray.io/market_type": "spot"}] + [{"ray.io/accelerator-type": "nvidia-h100"} * 2]
)
```

> Finally, we will implement a `fallback_strategy` API to support soft constraints or multiple deployment options if the initial `label_selector` cannot be satisfied.

Author: Done in [ed882e2](https://github.com/ray-project/enhancements/pull/60/commits/ed882e220ac3c02ef5a7f5ccc66312a2342d072c)

```python
@ray.remote(
    label_selector={"instance_type": "m5.16xlarge"},
    fallback_strategy=[
        # Fall back to an empty set of labels (no constraints).
        # This is equivalent to a "soft" constraint for an m5.16xlarge.
        {"label_selector": {}},
    ],
)
```

> ### Implementation plan

Author: Added an item in https://github.com/ray-project/ray/issues/51564 for this

Overview of Ray scheduler steps during label based scheduling:
1. Ray gets a request to schedule nodes based on some resources and labels.
2. Ray filters the feasible nodes by those that satisfy the resource request. A feasible node is one with sufficient total resources to satisfy the request, although those resources may not currently be available.  
3. Ray hard matches nodes that satisfy the resource request with those that satisfy the label selector and expression.
4. If no nodes match and a `fallback_strategy` is provided, filter by the provided fallback label selector and return the list of candidate nodes.
5. Ray returns the best schedulable node from the list of available (or feasible if no nodes are available) that satisfy the expressions in steps 3 and/or 4.

> Remaining steps to implement the label based scheduling feature:

Author: Sounds good, done in [0555f39](https://github.com/ray-project/enhancements/pull/60/commits/0555f39b347027ef9922c5d6ea4fbeb01fe2ca7b)

- [ ] (P1) Update `--labels` argument to take either a list of strings or read from file and expose this API publicly
- [ ] (P1) Add `label_selector` API to `@ray.remote` decorator to schedule tasks/actors
- [ ] (P1) Add `bundle_label_selector` to the `ray.util.placement_group` constructor to apply a set of `label_selector`s to placement group bundles
- [x] (P1) Ray saves label info associated with a node in `GcsNodeInfo` - already implemented
- [ ] (P1) Update `ClusterResourceScheduler::GetBestSchedulableNode` to enforce `label_selector` conditions when returning list of candidate nodes. This will eventually replace `SchedulingOptions::NodeLabelScheduling(scheduling_strategy)`.
- [ ] (P1) Populate list of default labels automatically, currently only supports `ray.io/node-id`
- [ ] (P1) Adapt Ray V2 Autoscaler to parse labels from K8s Pod Spec and generate a `--labels` arg to `rayStartParams`
- [ ] (P1) Update Autoscaler data model to pass label information by adding a labels field to the `ResourceRequest` message
- [ ] (P1) Update Autoscaler bin packing logic to directly consider label matching

> - [ ] (P2) Determine whitelist of K8s labels to always pass to Ray nodes

Reviewer 3: We probably want a way to add additional labels via RayCluster resource in KubeRay. It should support adding labels for the head pod and each worker group

Author: Added an item in https://github.com/ray-project/ray/issues/51564

### Should this change be within `ray` or outside?

The change should be within Ray since it's a direct enhancement to the Ray scheduler.

## Stewardship
### Required Reviewers
@MengjinYan

### Shepherd of the Proposal (should be a senior committer)
@edoakes

## Design and Architecture

### Current implementation state

Ray currently supports passing labels to a node through `ray start` with the `--labels` flag in Python, parsing labels from a json string with `parse_node_labels_json`, and associates a list of labels with a Node through the `_labels` field. Node information, including labels, are saved in the `GcsNodeInfo` data struct when a node is added. Ray also supports setting default labels on node add, but currently only sets `ray.io/node-id`.

To pass labels to a Ray node:
```sh
ray start --head --labels='{"gpu_type": "A100", "region": "us"}'
```

To access node labels:
```python

> ray.nodes()[0]["Labels"] == {"gpu_type": "A100", "region": "us"}

Reviewer 3: nit: use ray.io/accelerator-type instead of gpu_type in all the examples

Author: [ed882e2](https://github.com/ray-project/enhancements/pull/60/commits/ed882e220ac3c02ef5a7f5ccc66312a2342d072c)

    const Node &node, const rpc::LabelMatchExpression &expression) const {
    # Returns a bool based on whether a node's labels satisfy the given condition.
    # Supports exists, not exists, in, and not in conditions. We should also extend 
    # support to equal and not equal.
}
```

### Brief idea
In order to implement full label based scheduling as described in the [public proposal](https://docs.google.com/document/d/1DKhPuZERlLbsU4TIQZZ6POCsm1pVMBgN_yn5r0OmaDI), there are several required changes to the existing API and internal implementation in Ray core. Since most of the core functionality for adding, storing, and retrieving node labels is already implemented, the primary changes proposed here are to update the APIs, support autoscaling, and directly schedule nodes based on label selectors when provided, rather than passing a separate scheduling policy.

### API Design

To pass labels to a Ray node, we will amend the `--labels` argument to `ray start` to accept a string list of key-value pairs. Currently the labels argument accepts a json struct.
```sh
ray start --labels "key1=val1,key2=val2"
```

We will also support sourcing labels from a file using bash:
```sh

> ray start --labels $(cat labels.txt)

Reviewer 3: If we want to make it easy to pass down Pod labels into Raylet using downward API, we may want to support a separate flag `--labels-from-file` that reads labels in YAML format. This is because downward API for `metadata.labels` writes in YAML format WDYT @MengjinYan @edoakes

Reviewer: does the bash command to cat a file not work for this case? (I'm not familiar with downward API)

Reviewer: oh is the problem that the format will be YAML which doesn't match the comma-delimited string?

Reviewer: I'm fine to add a `--labels-file` to support that case; we can support `YAML` (which also supports natively includes`JSON`).

Do we need composability between the `--labels` and the file? Or are they mutually exclusive?

Author: It would make sense to me for them to be composable, so that the user could specify additional labels with `--labels` manually that may not be included in the labels file.

Author: Added this in [eb2aab3](https://github.com/ray-project/enhancements/pull/60/commits/eb2aab3b527e9e66083a7fc18afaeae6c8964b83)

### API Design

To pass labels to a Ray node, we will amend the `--labels` argument to `ray start` and `ray init` to accept a string list of key-value pairs. Currently the labels argument accepts a json struct.
```sh
ray start --labels "key1=val1,key2=val2"

# or
ray.init(labels="key1=val1,key2=val2")
```

We will also support sourcing labels from a file using bash for `ray start` only. This command will read labels in YAML format to support passing down Pod labels into the Raylet using downward API. The labels passed in from file should be composable with those specified by `--labels`.
```sh
ray start --labels-from-file $(cat labels.txt)
```

> To then schedule based on these labels, we will support passing a `label_selector` argument to the `@ray.remote` decorator. Adding this API here, rather than as a task/actor `scheduling_strategy`, will enable users to utilize label selectors in addition to other scheduling strategies.

Reviewer 2: Note that we will support the `label_selector` in both the `@ray.remote` decorator and the ray task/actor options.

Author: [ecc2967](https://github.com/ray-project/enhancements/pull/60/commits/ecc2967090e850660f12be92b28b0ede3d5d649c)

    const Node &node, const rpc::LabelMatchExpression &expression) const {
    # Returns a bool based on whether a node's labels satisfy the given condition.
    # Supports exists, not exists, in, and not in conditions. We should also extend 
    # support to equal and not equal.
}
```

### Brief idea
In order to implement full label based scheduling as described in the [public proposal](https://docs.google.com/document/d/1DKhPuZERlLbsU4TIQZZ6POCsm1pVMBgN_yn5r0OmaDI), there are several required changes to the existing API and internal implementation in Ray core. Since most of the core functionality for adding, storing, and retrieving node labels is already implemented, the primary changes proposed here are to update the APIs, support autoscaling, and directly schedule nodes based on label selectors passed to Ray tasks/actors, rather than requiring a separate scheduling policy.

### API Design

To pass labels to a Ray node, we will amend the `--labels` argument to `ray start` and `ray init` to accept a string list of key-value pairs. Currently the labels argument accepts a json struct.
```sh
ray start --labels "key1=val1,key2=val2"

# or

> ray.init(labels="key1=val1,key2=val2")

Reviewer: `ray.init` should be made to be a Pythonic API, so this should take a `Dict[str,str]` rather than a string

Author: Oh oops, done in [9ab94a2](https://github.com/ray-project/enhancements/pull/60/commits/9ab94a251764f9f0111b35b4dbfa68ae2dff74d8)

```

> We will also support sourcing labels from a file using bash for `ray start` only. This command will read labels in YAML format to support passing down Pod labels into the Raylet using downward API. The labels passed in from file should be composable with those specified by `--labels`.

Reviewer: What does `composable` mean? Which takes precedence if there is a conflict? In my opinion, should be CLI arg >> file.

Author: Yeah that's what I meant by composable, and sounds good. I implemented it with the CLI arg taking precedence over the file arg in this PR: https://github.com/ray-project/ray/pull/51706/files#

Author: Clarified this in [9ab94a2](https://github.com/ray-project/enhancements/pull/60/commits/9ab94a251764f9f0111b35b4dbfa68ae2dff74d8).

Reviewer: perfect 👍

```sh

> ray start --labels-from-file $(cat labels.txt)

Reviewer: ```suggestion
ray start --labels-file $(cat labels.txt)
```

## Summary
### General Motivation

This REP summarizes the current state of the node label scheduling feature enhancement and the remaining work to fully support scheduling using label selectors in Ray. This REP supersedes the previous [node affinity feature enhancement REP](https://github.com/ryanaoleary/enhancements/blob/main/reps/2023-02-03-node-affinity-feature-enhancements.md).

### Should this change be within `ray` or outside?

The change should be within Ray since it's a direct enhancement to the Ray scheduler.

## Stewardship
### Required Reviewers

> @MengjinYan

Reviewer 3: can you add me as a reviewer?

Author: [66362f4](https://github.com/ray-project/enhancements/pull/60/commits/66362f4e4effc48beb67bebca88f537425fa1d84)

    const Node &node, const rpc::LabelMatchExpression &expression) const {
    # Returns a bool based on whether a node's labels satisfy the given condition.
    # Supports exists, not exists, in, and not in conditions. We should also extend 
    # support to equal and not equal.
}
```

### Brief idea
In order to implement full label based scheduling as described in the [public proposal](https://docs.google.com/document/d/1DKhPuZERlLbsU4TIQZZ6POCsm1pVMBgN_yn5r0OmaDI), there are several required changes to the existing API and internal implementation in Ray core. Since most of the core functionality for adding, storing, and retrieving node labels is already implemented, the primary changes proposed here are to update the APIs, support autoscaling, and directly schedule nodes based on label selectors passed to Ray tasks/actors, rather than requiring a separate scheduling policy.

### API Design

To pass labels to a Ray node, we will amend the `--labels` argument to `ray start` and `ray init` to accept a string list of key-value pairs. Currently the labels argument accepts a json struct.
```sh
ray start --labels "key1=val1,key2=val2"
```

Or with `ray init`:
```python

> ray.init(labels={"key1": "val1", "key2": "val2"})

Author: Oh oops, I'll take this segment out and add the note about the warning.

> ray.init(labels={"key1": "val1", "key2": "val2"})

Author: [59e6c5d](https://github.com/ray-project/enhancements/pull/60/commits/59e6c5d4dfa816352dac60368e56546a607b19d0)

```

We will also support sourcing labels from a file using bash for `ray start` only. This command will read labels in YAML format to support passing down Pod labels into the Raylet using downward API. The labels passed in from file should be composable with those specified by `--labels`, with the value in `--labels` taking precedence if there is a conflict.
```sh

> ray start --labels-file $(cat labels.txt)

Reviewer 3: Should this be:
```
ray start --labels-file labels.txt
```

Or more specifically it should be an absolute path
```
ray start --labels-file /path/to/labels.yaml
```

Author: I started implementing it here: https://github.com/ray-project/ray/pull/51706/files as `--labels-from-file` (I need to update the name) where it just sources the contents of the file with `$(cat labels.txt)`. If we want to read directly from a file instead I can change it. Should we support both txt and yaml files (I guess both would be expected in yaml format)?

Reviewer: If we just want to source from a file with `cat labels.txt` we don't need to add anything inside of Ray -- bash does the trick :)

Andrew mentioned it'd be better to support reading YAML files for the "downward" API, that's the impetus for supporting `--labels-file` explicitly.

Reviewer 3: yeah agreed, `$(cat labels.txt)` is perfectly fine, but you would pass that into `ray start --labels`, not `--labels-file` was my point.

Reviewer 2: So if I understand correctly, for the string input, no matter if it is specified in the command or if it is sourcing from a file, we want to use the `--labels`. And in both cases, we expects the input format to be `key1=value1,key2=value2`. 

And at the same time, if we want to support using "downward" API with yamal file as well, the idea is either to have a `--labels-file` option and pass in the yaml file name so that `ray start` can read the file directory, or more simply, just use the `--labels` options and add the file read & translation logic outside of the ray logic. 

And I believe we can always start from the simpler solution. Is that right?

Author: Ahh I see, I was expecting `--labels-file` to source from the file but just to expect the string in yaml format (so both would options would take a string but then parse them from different formats). I updated the REP to reflect reading directly from a file in [3c444e7](https://github.com/ray-project/enhancements/pull/60/commits/3c444e71e87ca9582ab92264a112c8193f394f49). I'll update  https://github.com/ray-project/ray/pull/51706 to that implementation as well.
